Yoast/wordpress-seo
# Fix SEMrush authentication flow

## Context


* If the user is not logged in SEMrush (= no valid tokens are stored in the DB), a popup opens when clicking on the button to get the related keyphrases. We should manage the whole authorization flow from this point on.

## Summary


This PR can be summarized in the following changelog entry:

* Fixes the authorization flow within the SEMrush integration

## Relevant technical choices:

* The PR adds a check on the keyphrase (not present in P3-57 #15927 where the popup was originally added) to prevent starting the auth flow if the keyphrase is empty. While a keyphrase is not needed to perform the authorization, it is needed to open the modal (and send the request to SEMrush) when the auth flow has been successful.
* The focus is set to the popup window after it has been opened. This could be useful if the window has slipped below other windows and the user clicks again on the link to open it once more. Note that the flow starts from the beginning (= the location is set again to the starting OAuth URL) if the popup was opened from the metabox and "focused" from the sidebar, or vice-versa.
* At the moment only a successful flow provides meaningful feedback to the user (the popup closes and the modal opens). If the authentication fails, error are sent to the JS console. If the user denies the authorization with SEMrush, the popup just closes.
* This PR also includes a change that was overlooked in P3-57 (#15927): when the SEMrush integration is disabled (via the toggle in the plugin's settings) no check for token existence or validity is performed (thus avoiding unnecessary operations and remote calls to SEMrush API).

## Test instructions

This PR can be tested by following these steps:

### :exclamation:  How to reset the auth tokens in the database (needed to perform the tests below)
* With a DB inspector, search for the `wpseo` array of options in `wp_options` table
* make sure you copy the content of the option elsewhere for backup
* edit the serialized array so the part which looks like:
```
s:14:"semrush_tokens";a:5:{s:12:"access_token";s:40:"xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx";s:13:"refresh_token";s:40:"xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx";s:7:"expires";i:1598357188;s:11:"has_expired";b:0;s:10:"created_at";i:1597752388;}
```
becomes
```
s:14:"semrush_tokens";a:0:{}
```

[Edit by Andrea]: make sure you're not logged in to semrush.com before following the test steps

### Test the successful flow
* make sure you have credentials for SEMrush, as registration make things much more difficult for the moment and will be better addressed later
* reset the tokens as detailed above
* edit a (new) post
* do not enter a keyphrase, click the "Get related keyphrases" button and see the error message
* enter a keyphrase, click the "Get related keyphrases" button and see the popup opens
* Inspect the Redux state for `yoast-seo/editor`, see that `root` > `SEMrushRequest` > `isLoggedIn` is `false`
* login into SEMrush (you are not already) and click on "Approve"
* see that the popup closes, `isLoggedIn` becomes `true` and the modal opens displaying the results for your keyphrase.
* inspect `wpseo` in the DB to see that tokens have been stored in it.
* repeat the check with Classic Editor

### Test the denial of authorization by the user
* reset the tokens as detailed above
* edit a (new) post
* enter a keyphrase, click the "Get related keyphrases" button and see the popup opens
* Inspect the Redux state for `yoast-seo/editor`, see that `root` > `SEMrushRequest` > `isLoggedIn` is `false`
* login into SEMrush (you are not already) and click on "Deny"
* see that the popup closes and `isLoggedIn` stays `false`.
* see that clicking the "Get related keyphrases" button the popup opens again, asking again for authorization.
* inspect `wpseo` in the DB to see that no tokens have been added.
* repeat the check with Classic Editor

### Test errors (optional)

We might have a number of possible problems after the authorization is granted by the user on the popup, e.g.
* the message from SEMrush has no URL or a bad URL
* the message from SEMrush has a URL but no `code` parameter can be extracted from it
* the authentication via the Yoast API fails
* etc.

All these things are difficult (even impossible) to reproduce "by hand" (see automatic tests).
A possible way to check that the flow stops and the console shows related error messages could be:

* reset the tokes as detailed above
* either: 
  * change ` js/src/components/SEMrushRelatedKeyphrasesModal.js` at the line 121 to pass a malformed url to `URL()`
  * change ` js/src/components/SEMrushRelatedKeyphrasesModal.js` at the line 121 to pass `URL()` an url without a `code` in the query string
  * note by Andrea: the line number became 139 in the meantime and may change in the future: for better reference, this is the `new URL()` within the `performAuthenticationRequest()` function
  * change ` js/src/components/SEMrushRelatedKeyphrasesModal.js` at the line 126 to set `code` in the `data` object to a non-valid code (ay string will work)
  * note by Andrea: the line number became 144 in the meantime and may change in the future: for better reference, this is the `data: { code: code }` within the `performAuthenticationRequest()` function
* build
* open the popup, "Accept" the authentication, see that the modal closes, the browser JS console shows an error related to the problem you inserted in the code, check that the "Get related keyphrases" button still opens the popup and the `wpseo` option does not contain tokens.

### Test that the refreshing of the token is disabled if the SEMrush integration is disabled
* In the `wpseo` array of options, do not remove the tokens, but change `"has_expired";b:0;` to `"has_expired";b:1;`
* Edit a (new) post
* Inspect again the `wpseo` array and see that it contains `"has_expired";b:0;` again. This means that the token has been refreshed with SEMrush since it was marked as expired.
* Visit SEO→General, tab "Integrations", disable the SEMrush integration and save.
* In the `wpseo` array of options, change `"has_expired";b:0;` to `"has_expired";b:1;`
* Edit a (new) post
* Inspect again the `wpseo` array and see that it still contains `"has_expired";b:1;`. This means that the token has **not** been refreshed with SEMrush.

### Unit tests
The PR adds tests for the features that have been added to the `SEMrushRelatedKeyphrasesModal` component, so
* run `yarn test`
* also check Travis' first job is OK

## UI changes

## Documentation

## Quality assurance


Fixes [P3-54]
